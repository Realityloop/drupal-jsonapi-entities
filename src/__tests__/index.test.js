import axios from 'axios'
import drupalJSONAPIEntities from '..'

import mockEntityFormDisplay from '../resources/__fixtures__/entity_form_display'
import mockEntityViewDisplay from '../resources/__fixtures__/entity_view_display'
import mockFieldConfig from '../resources/__fixtures__/field_config'
import mockFieldStorageConfig from '../resources/__fixtures__/field_storage_config'

// Setup mock axios module.
jest.mock('axios')
const axiosMock = mockData => {
  axios.create.mockReturnValue({
    interceptors: { request: { use: () => { } } },
    post: () => ({ data: mockData })
  })
}

const baseURL = 'https://example.com'

describe('drupalJSONAPIEntities', () => {
  test('constructor', () => {
    axiosMock({})

    // Throw error if 'baseURL' not provided.
    expect(() => { new drupalJSONAPIEntities() }).toThrow('The \'baseURL\' parameter is required.')

    // Ensure class type.
    expect(new drupalJSONAPIEntities(baseURL)).toBeInstanceOf(drupalJSONAPIEntities)
  })

  test('getFormSchema', async () => {
    axiosMock({
      'node--recipe--default--entity_form_display': { body: JSON.stringify(mockEntityFormDisplay) },
      'node--recipe--field_config': { body: JSON.stringify(mockFieldConfig) },
      'node--field_storage_config': { body: JSON.stringify(mockFieldStorageConfig) }
    })
    const instance = new drupalJSONAPIEntities(baseURL)

    const schema = await instance.getFormSchema('node', 'recipe')
    expect(schema.fields.items).toMatchSnapshot()
    expect(schema.groups.items).toMatchSnapshot()
    expect(schema.keys).toMatchSnapshot()
  })

  test('getViewSchema', async () => {
    axiosMock({
      'node--recipe--default--entity_view_display': { body: JSON.stringify(mockEntityViewDisplay) },
      'node--recipe--field_config': { body: JSON.stringify(mockFieldConfig) }
    })
    const instance = new drupalJSONAPIEntities(baseURL)

    const schema = await instance.getViewSchema('node', 'recipe')
    expect(schema.fields.items).toMatchSnapshot()
    expect(schema.groups.items).toMatchSnapshot()
    expect(schema.keys).toMatchSnapshot()
  })
})
