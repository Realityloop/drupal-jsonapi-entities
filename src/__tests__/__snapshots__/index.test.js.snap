// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`drupalJSONAPIEntities getFormSchema 1`] = `
Object {
  "created": Object {
    "id": "created",
    "property": true,
    "settings": Object {},
    "type": "datetime_timestamp",
    "weight": 13,
  },
  "field_author": Object {
    "cardinality": 1,
    "description": "The author of this recipe. Start typing their username to search.",
    "id": "field_author",
    "label": "Author",
    "property": false,
    "required": true,
    "settings": Object {
      "handler": "default:user",
      "handler_settings": Object {
        "auto_create": false,
        "filter": Object {
          "type": "_none",
        },
        "include_anonymous": true,
        "sort": Object {
          "field": "_none",
        },
        "target_bundles": null,
      },
      "match_operator": "CONTAINS",
      "placeholder": "",
      "size": 60,
      "target_type": "user",
    },
    "type": "entity_reference_autocomplete",
    "weight": 1,
  },
  "field_cooking_time": Object {
    "cardinality": 1,
    "description": "",
    "group": "group_meta",
    "id": "field_cooking_time",
    "label": "Cooking time",
    "property": false,
    "required": false,
    "settings": Object {
      "max": null,
      "min": 0,
      "placeholder": "",
      "prefix": "",
      "size": "normal",
      "suffix": " minutes",
      "unsigned": false,
    },
    "type": "number",
    "weight": 3,
  },
  "field_difficulty": Object {
    "cardinality": 1,
    "description": "",
    "id": "field_difficulty",
    "label": "Difficulty",
    "property": false,
    "required": true,
    "settings": Object {
      "allowed_values": Object {
        "easy": "Easy",
        "hard": "Hard",
        "medium": "Medium",
      },
      "allowed_values_function": "",
    },
    "type": "options_select",
    "weight": 5,
  },
  "field_image": Object {
    "cardinality": 1,
    "description": "",
    "id": "field_image",
    "label": "Image",
    "property": false,
    "required": false,
    "settings": Object {
      "handler": "default:media",
      "handler_settings": Object {
        "auto_create": false,
        "auto_create_bundle": "",
        "sort": Object {
          "field": "_none",
        },
        "target_bundles": Object {
          "image": "image",
        },
      },
      "media_types": Array [],
      "target_type": "media",
    },
    "type": "media_library_widget",
    "weight": 8,
  },
  "field_ingredients": Object {
    "cardinality": -1,
    "description": "List the ingredients required for this recipe, one per item.",
    "id": "field_ingredients",
    "label": "Ingredients",
    "property": false,
    "required": false,
    "settings": Object {
      "case_sensitive": false,
      "is_ascii": false,
      "max_length": 255,
      "placeholder": "",
      "size": 60,
    },
    "type": "string_textfield",
    "weight": 10,
  },
  "field_number_of_servings": Object {
    "cardinality": 1,
    "description": "",
    "id": "field_number_of_servings",
    "label": "Number of servings",
    "property": false,
    "required": true,
    "settings": Object {
      "max": null,
      "min": 0,
      "placeholder": "",
      "prefix": "",
      "size": "normal",
      "suffix": "",
      "unsigned": false,
    },
    "type": "number",
    "weight": 4,
  },
  "field_preparation_time": Object {
    "cardinality": 1,
    "description": "",
    "group": "group_meta",
    "id": "field_preparation_time",
    "label": "Preparation time",
    "property": false,
    "required": true,
    "settings": Object {
      "max": null,
      "min": 0,
      "placeholder": "",
      "prefix": "",
      "size": "normal",
      "suffix": " minutes",
      "unsigned": false,
    },
    "type": "number",
    "weight": 2,
  },
  "field_recipe_category": Object {
    "cardinality": -1,
    "description": "",
    "id": "field_recipe_category",
    "label": "Recipe category",
    "property": false,
    "required": false,
    "settings": Object {
      "handler": "default:taxonomy_term",
      "handler_settings": Object {
        "auto_create": false,
        "auto_create_bundle": "",
        "sort": Object {
          "direction": "asc",
          "field": "name",
        },
        "target_bundles": Object {
          "recipe_category": "recipe_category",
        },
      },
      "target_type": "taxonomy_term",
    },
    "type": "options_select",
    "weight": 6,
  },
  "field_recipe_instruction": Object {
    "cardinality": 1,
    "description": "",
    "id": "field_recipe_instruction",
    "label": "Recipe instruction",
    "property": false,
    "required": true,
    "settings": Object {
      "placeholder": "",
      "rows": 5,
    },
    "type": "text_textarea",
    "weight": 11,
  },
  "field_summary": Object {
    "cardinality": 1,
    "description": "Provide a short overview of this recipe.",
    "id": "field_summary",
    "label": "Summary",
    "property": false,
    "required": true,
    "settings": Object {
      "placeholder": "",
      "rows": 5,
    },
    "type": "text_textarea",
    "weight": 9,
  },
  "field_tags": Object {
    "cardinality": -1,
    "description": "",
    "id": "field_tags",
    "label": "Tags",
    "property": false,
    "required": false,
    "settings": Object {
      "handler": "default:taxonomy_term",
      "handler_settings": Object {
        "auto_create": false,
        "auto_create_bundle": "",
        "sort": Object {
          "direction": "asc",
          "field": "name",
        },
        "target_bundles": Object {
          "tags": "tags",
        },
      },
      "match_operator": "CONTAINS",
      "placeholder": "",
      "size": 60,
      "target_type": "taxonomy_term",
    },
    "type": "entity_reference_autocomplete_tags",
    "weight": 7,
  },
  "langcode": Object {
    "id": "langcode",
    "property": true,
    "settings": Object {
      "include_locked": true,
    },
    "type": "language_select",
    "weight": 2,
  },
  "path": Object {
    "id": "path",
    "property": true,
    "settings": Object {},
    "type": "path",
    "weight": 16,
  },
  "promote": Object {
    "id": "promote",
    "property": true,
    "settings": Object {
      "display_label": true,
    },
    "type": "boolean_checkbox",
    "weight": 14,
  },
  "status": Object {
    "id": "status",
    "property": true,
    "settings": Object {
      "display_label": true,
    },
    "type": "boolean_checkbox",
    "weight": 17,
  },
  "sticky": Object {
    "id": "sticky",
    "property": true,
    "settings": Object {
      "display_label": true,
    },
    "type": "boolean_checkbox",
    "weight": 15,
  },
  "title": Object {
    "id": "title",
    "property": true,
    "settings": Object {
      "placeholder": "",
      "size": 60,
    },
    "type": "string_textfield",
    "weight": 0,
  },
  "uid": Object {
    "id": "uid",
    "property": true,
    "settings": Object {
      "match_operator": "CONTAINS",
      "placeholder": "",
      "size": 60,
    },
    "type": "entity_reference_autocomplete",
    "weight": 12,
  },
}
`;

exports[`drupalJSONAPIEntities getFormSchema 2`] = `
Array [
  "created",
  "field_author",
  "field_cooking_time",
  "field_difficulty",
  "field_image",
  "field_ingredients",
  "field_number_of_servings",
  "field_preparation_time",
  "field_recipe_category",
  "field_recipe_instruction",
  "field_summary",
  "field_tags",
  "langcode",
  "path",
  "promote",
  "status",
  "sticky",
  "title",
  "uid",
]
`;

exports[`drupalJSONAPIEntities getViewSchema 1`] = `
Object {
  "field_cooking_time": Object {
    "description": "",
    "group": "group_meta",
    "id": "field_cooking_time",
    "label": "Cooking time",
    "labelPosition": "above",
    "property": false,
    "required": false,
    "settings": Object {
      "max": null,
      "min": 0,
      "prefix": "",
      "prefix_suffix": true,
      "suffix": " minutes",
      "thousand_separator": "",
    },
    "thirdPartySettings": Array [],
    "type": "number_integer",
    "weight": 1,
  },
  "field_difficulty": Object {
    "description": "",
    "id": "field_difficulty",
    "label": "Difficulty",
    "labelPosition": "above",
    "property": false,
    "required": true,
    "settings": Object {},
    "thirdPartySettings": Array [],
    "type": "list_default",
    "weight": 2,
  },
  "field_image": Object {
    "description": "",
    "id": "field_image",
    "label": "Image",
    "labelPosition": "hidden",
    "property": false,
    "required": false,
    "settings": Object {
      "handler": "default:media",
      "handler_settings": Object {
        "auto_create": false,
        "auto_create_bundle": "",
        "sort": Object {
          "field": "_none",
        },
        "target_bundles": Object {
          "image": "image",
        },
      },
      "image_link": "",
      "responsive_image_style": "3_2_image",
    },
    "thirdPartySettings": Array [],
    "type": "responsive_image",
    "weight": 3,
  },
  "field_ingredients": Object {
    "description": "List the ingredients required for this recipe, one per item.",
    "id": "field_ingredients",
    "label": "Ingredients",
    "labelPosition": "above",
    "property": false,
    "required": false,
    "settings": Object {
      "link_to_entity": false,
    },
    "thirdPartySettings": Array [],
    "type": "string",
    "weight": 4,
  },
  "field_number_of_servings": Object {
    "description": "",
    "id": "field_number_of_servings",
    "label": "Number of servings",
    "labelPosition": "above",
    "property": false,
    "required": true,
    "settings": Object {
      "max": null,
      "min": 0,
      "prefix": "",
      "prefix_suffix": true,
      "suffix": "",
      "thousand_separator": "",
    },
    "thirdPartySettings": Array [],
    "type": "number_integer",
    "weight": 5,
  },
  "field_preparation_time": Object {
    "description": "",
    "group": "group_meta",
    "id": "field_preparation_time",
    "label": "Preparation time",
    "labelPosition": "above",
    "property": false,
    "required": true,
    "settings": Object {
      "max": null,
      "min": 0,
      "prefix": "",
      "prefix_suffix": true,
      "suffix": " minutes",
      "thousand_separator": "",
    },
    "thirdPartySettings": Array [],
    "type": "number_integer",
    "weight": 6,
  },
  "field_recipe_category": Object {
    "description": "",
    "id": "field_recipe_category",
    "label": "Recipe category",
    "labelPosition": "above",
    "property": false,
    "required": false,
    "settings": Object {
      "handler": "default:taxonomy_term",
      "handler_settings": Object {
        "auto_create": false,
        "auto_create_bundle": "",
        "sort": Object {
          "direction": "asc",
          "field": "name",
        },
        "target_bundles": Object {
          "recipe_category": "recipe_category",
        },
      },
      "link": true,
    },
    "thirdPartySettings": Array [],
    "type": "entity_reference_label",
    "weight": 7,
  },
  "field_recipe_instruction": Object {
    "description": "",
    "id": "field_recipe_instruction",
    "label": "Recipe instruction",
    "labelPosition": "above",
    "property": false,
    "required": true,
    "settings": Object {},
    "thirdPartySettings": Array [],
    "type": "text_default",
    "weight": 8,
  },
  "field_summary": Object {
    "description": "Provide a short overview of this recipe.",
    "id": "field_summary",
    "label": "Summary",
    "labelPosition": "hidden",
    "property": false,
    "required": true,
    "settings": Object {},
    "thirdPartySettings": Array [],
    "type": "text_default",
    "weight": 9,
  },
  "field_tags": Object {
    "description": "",
    "id": "field_tags",
    "label": "Tags",
    "labelPosition": "above",
    "property": false,
    "required": false,
    "settings": Object {
      "handler": "default:taxonomy_term",
      "handler_settings": Object {
        "auto_create": false,
        "auto_create_bundle": "",
        "sort": Object {
          "direction": "asc",
          "field": "name",
        },
        "target_bundles": Object {
          "tags": "tags",
        },
      },
      "link": true,
    },
    "thirdPartySettings": Array [],
    "type": "entity_reference_label",
    "weight": 10,
  },
  "links": Object {
    "id": "links",
    "labelPosition": undefined,
    "settings": Object {},
    "thirdPartySettings": Array [],
    "type": undefined,
    "weight": 0,
  },
  "title": Object {
    "id": "title",
    "labelPosition": "hidden",
    "settings": Object {
      "link_to_entity": false,
    },
    "thirdPartySettings": Array [],
    "type": "string",
    "weight": -5,
  },
}
`;

exports[`drupalJSONAPIEntities getViewSchema 2`] = `
Array [
  "field_cooking_time",
  "field_difficulty",
  "field_image",
  "field_ingredients",
  "field_number_of_servings",
  "field_preparation_time",
  "field_recipe_category",
  "field_recipe_instruction",
  "field_summary",
  "field_tags",
  "links",
  "title",
]
`;
